packages/osd-dev-utils/src/proc_runner/proc.ts

```diff
@@ -153,15 +153,15 @@ export function startProc(name: string, options: ProcOptions, log: ToolingLog) {
     await withTimeout(
       async () => {
         log.debug(`Sending "${signal}" to proc "${name}"`);
-        await treeKillAsync(childProcess.pid, signal);
+        await treeKillAsync(childProcess.pid!, signal);
```

Comment:
This fix is due to this error
```
ERROR Error: Command failed with exit code 2: /usr/share/yarn/bin/yarn.js run osd:bootstrap
      error Command failed with exit code 2.
      error Command failed with exit code 2.
      $ yarn build
      $ tsc
      src/proc_runner/proc.ts(156,29): error TS2345: Argument of type 'number | undefined' is not assignable to parameter of type 'number'.
        Type 'undefined' is not assignable to type 'number'.
      info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
      info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
          at makeError (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:25045:11)
          at handlePromise (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:23981:26)
          at processTicksAndRejections (node:internal/process/task_queues:96:5)
          at async /home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:9238:9
          at async scheduleItem (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:11285:9)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
The problem is with the `childProcess.pid` value. In Node.js, the pid is undefined if the child process has not been successfully spawned (or if the child process has already exited). Hence, `childProcess.pid` is of type `number | undefined`. The `treeKillAsync` function expects a PID of type number, not `number | undefined`. Therefore, we see ts throws an error.

`!`  allows ts to treat `childProcess.pid` as type number rather than `number | undefined`. An optional chain `childProcess.pid?` would not solve the issue because it would still result in undefined if pid doesn't exist, and treeKillAsync requires a number. Here is the result of using `?`:

```
ubuntu@ip-172-31-55-237:~/OpenSearch-Dashboards$ yarn osd bootstrap
yarn run v1.22.19
$ scripts/use_node scripts/osd bootstrap
 info [opensearch-dashboards] running yarn

$ scripts/use_node ./preinstall_check
[1/4] Resolving packages...
warning Resolution field "typescript@4.0.2" is incompatible with requested version "typescript@~4.5.2"
success Already up-to-date.

 succ yarn.lock analysis completed without any issues
 succ 8 bootstrap builds are cached
 info [@osd/dev-utils] running [osd:bootstrap] script
ERROR [bootstrap] failed:
ERROR Error: Command failed with exit code 2: /usr/share/yarn/bin/yarn.js run osd:bootstrap
      error Command failed with exit code 2.
      error Command failed with exit code 2.
      $ yarn build
      $ tsc
      src/proc_runner/proc.ts(156,46): error TS1109: Expression expected.
      info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
      info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
          at makeError (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:25045:11)
          at handlePromise (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:23981:26)
          at processTicksAndRejections (node:internal/process/task_queues:96:5)
          at async /home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:9238:9
          at async scheduleItem (/home/ubuntu/OpenSearch-Dashboards/packages/osd-pm/dist/index.js:11285:9)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
``

Reply:
Right - I was suggesting that what you really needed here was a optional chaining with a fallback value so that you never pass `undefined` to `treeKillAsync`. The non-null assertion doesn't do anything other than tell typescript not to worry about the potential of `undefined` - but if `pid` is undefined, you'll still pass that to a function that isn't prepared to handle that. What you probably want instead is a type guard, so that you only call `treeKillAsync` when `pid` is defined.